WWW::Mechanized::Cache warns on explicit cache

WWW::Mechanize::Cache emits a warning when `cache` is explicitly provided. This can be seen running the installing test suite (see end of the comment)

This is because arguments that LWP::UserAgent does not expect are being passed directly to it. This can be easily fixed by adding the following anywhere to the module:

```
+ sub FOREIGNBUILDARGS {
+    my ($class, %args) = @_;
+    delete $args{cache};
+    delete $args{_verbose_dwarn};
+    return %args;
+}
+
```

It appears that `t/002-bad-custom-cache.t` is already testing this (though it may wish to check for `_verbose_dwarn` behavior as well):

```
~/programming/tmp/WWW-Mechanize-Cached-1.51 $ make test
cp lib/WWW/Mechanize/Cached.pm blib/lib/WWW/Mechanize/Cached.pm
PERL_DL_NONLAZY=1 /usr/bin/perl "-MExtUtils::Command::MM" "-e" "test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/00-report-prereqs.t ..... # 
# Versions for all modules listed in MYMETA.json (including optional ones):
# 
# === Configure Requires ===
# 
#     Module              Want Have
#     ------------------- ---- ----
#     ExtUtils::MakeMaker  any 6.66
# 
# === Build Requires ===
# 
#     Module              Want Have
#     ------------------- ---- ----
#     ExtUtils::MakeMaker  any 6.66
# 
# === Test Requires ===
# 
#     Module                 Want  Have
#     ---------------------- ---- -----
#     CHI                     any  0.60
#     Cache::FileCache        any undef
#     ExtUtils::MakeMaker     any  6.66
#     File::Spec              any  3.40
#     Find::Lib               any  1.04
#     HTTP::Request           any  6.00
#     Path::Class             any  0.33
#     Test::Fatal             any 0.013
#     Test::More              any  0.98
#     Test::Requires          any  0.07
#     Test::RequiresInternet  any  0.05
#     constant                any  1.27
#     lib                     any  0.63
# 
# === Test Recommends ===
# 
#     Module         Want     Have
#     ---------- -------- --------
#     CPAN::Meta 2.120900 2.150010
# 
# === Runtime Requires ===
# 
#     Module                           Want     Have
#     ---------------------------- -------- --------
#     Cache::FileCache                  any    undef
#     Carp                              any     1.29
#     Data::Dump                        any     1.22
#     Module::Runtime                   any    0.014
#     Moo                          1.004005 2.003001
#     MooX::Types::MooseLike::Base      any     0.29
#     Storable                         2.21     2.41
#     WWW::Mechanize                    any     1.87
#     namespace::clean                  any     0.24
#     strict                            any     1.07
#     warnings                          any     1.18
# 
# === Runtime Recommends ===
# 
#     Module Want Have
#     ------ ---- ----
#     CHI     any 0.60
# 
t/00-report-prereqs.t ..... ok   
t/002-bad-custom-cache.t .. 1/? 
#   Failed test 'Threw the right warning'
#   at t/002-bad-custom-cache.t line 21.
# found carped warning: Unrecognized LWP::UserAgent options: cache at t/002-bad-custom-cache.t line 16.
# found carped warning: The cache param must be an initialized cache object at t/002-bad-custom-cache.t line 16.
# expected to find warning: (?^:cache param)
# Looks like you failed 1 test of 2.
t/002-bad-custom-cache.t .. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/2 subtests  
t/003-basic.t ............. ok   
t/004-cached.t ............ 1/? Unrecognized LWP::UserAgent options: cache at t/004-cached.t line 48.
Unrecognized LWP::UserAgent options: cache at t/004-cached.t line 78.
t/004-cached.t ............ ok    
t/005-custom-cache.t ...... 1/? Unrecognized LWP::UserAgent options: _verbose_dwarn cache at t/005-custom-cache.t line 36.
t/005-custom-cache.t ...... ok   
t/006-cached-chi.t ........ 1/? Unrecognized LWP::UserAgent options: cache at t/006-cached-chi.t line 49.
Unrecognized LWP::UserAgent options: cache at t/006-cached-chi.t line 79.
t/006-cached-chi.t ........ ok    
t/cache_key.t ............. 1/? Unrecognized LWP::UserAgent options: cache at t/cache_key.t line 18.
# reversing page order
t/cache_key.t ............. ok    
t/cache_ok.t .............. ok   
t/default.t ............... ok   
t/is_cached.t ............. ok   

Test Summary Report
-------------------
t/002-bad-custom-cache.t (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=10, Tests=105,  2 wallclock secs ( 0.04 usr  0.02 sys +  1.51 cusr  0.19 csys =  1.76 CPU)
Result: FAIL
Failed 1/10 test programs. 1/105 subtests failed.
make: *** [test_dynamic] Error 255
```

--- a/Changes
+++ b/Changes
@@ -1,6 +1,7 @@
 Revision history for Perl module WWW::Mechanize::Cached
 
 {{$NEXT}}
+    - Remove warnings on new() (GH#15) (Jonathan Rubin, simbabque)
 
 1.52      2019-11-18 14:15:37Z
     - Implement invalidate_last_request to remove the last request from the

--- a/lib/WWW/Mechanize/Cached.pm
+++ b/lib/WWW/Mechanize/Cached.pm
@@ -34,6 +34,28 @@ has cache_mismatch_content_length => (
 
 has cache => ( is => 'lazy', isa => \&_isa_warn_cache );
 
+sub FOREIGNBUILDARGS {
+    my ($class, %args) = @_;
+
+    # WWW::Mechanize/LWP::UserAgent would complain about these
+    for my $attribute (
+        qw(
+            is_cached
+            positive_cache
+            ref_in_cach_key
+            _verbose_dwarn
+            cache_undef_content_length
+            cache_zero_content_length
+            cache_mismatch_content_length
+            cache
+        )
+    ) {
+        delete $args{ $attribute };
+    }
+
+    return %args;
+}
+
 sub _isa_warn_cache {
     return
             if 'HASH' ne ref $_[0]

--- a/t/007-initialize-warnings.t
+++ b/t/007-initialize-warnings.t
@@ -1,14 +1,9 @@
 use strict;
 use warnings FATAL => 'all';
+use Test::Warnings qw( :all :no_end_test );
 
 use Test::More;
 use WWW::Mechanize::Cached;
-
-BEGIN {
-    eval "use Test::Warn";
-    plan skip_all => "Test::Warn required for testing initialization warnings"
-        if $@;
-}
 
 use lib 't';
 use TestCache;
@@ -17,49 +12,40 @@ my $cache = TestCache->new();
 isa_ok( $cache, 'TestCache' );
 
 my $was_warning = $^W;
-   $^W          = 1;
+$^W = 1;
 
-my $mech;
-
-warnings_are
-    {
-        $mech = WWW::Mechanize::Cached->new(
-            cache => $cache,
-        );
-    }
-    [ ],
-    "No warnings for accepted 'cache' parameter";
+my $mech = WWW::Mechanize::Cached->new(
+    cache => $cache,
+);
+had_no_warnings("No warnings for accepted 'cache' parameter");
 
 for my $boolean_attribute (
     qw(
-        is_cached
-        positive_cache
-        ref_in_cach_key
-        _verbose_dwarn
-        cache_undef_content_length
-        cache_zero_content_length
-        cache_mismatch_content_length
+    is_cached
+    positive_cache
+    ref_in_cach_key
+    _verbose_dwarn
+    cache_undef_content_length
+    cache_zero_content_length
+    cache_mismatch_content_length
     )
 ) {
-    warnings_are
-        {
-            $mech = WWW::Mechanize::Cached->new(
-                $boolean_attribute => 1,
-            );
-        }
-        [ ],
-        "No warnings for accepted '$boolean_attribute' parameter";
+    $mech = WWW::Mechanize::Cached->new(
+        $boolean_attribute => 1,
+    );
+    had_no_warnings(
+        "No warnings for accepted '$boolean_attribute' parameter");
 }
 
-warnings_exist
-    {
+like(
+    warning {
         $mech = WWW::Mechanize::Cached->new(
             not_my_argument => 1,
         );
-    }
+    },
     qr/not_my_argument/,
-    'Unrecognized arguments passed through to WWW::Mechanize';
-
+    'Unrecognized arguments passed through to WWW::Mechanize'
+);
 
 # Put this back
 $^W = $was_warning;

--- a/t/007-initialize-warnings.t
+++ b/t/007-initialize-warnings.t
@@ -1,0 +1,67 @@
+use strict;
+use warnings FATAL => 'all';
+
+use Test::More;
+use WWW::Mechanize::Cached;
+
+BEGIN {
+    eval "use Test::Warn";
+    plan skip_all => "Test::Warn required for testing initialization warnings"
+        if $@;
+}
+
+use lib 't';
+use TestCache;
+
+my $cache = TestCache->new();
+isa_ok( $cache, 'TestCache' );
+
+my $was_warning = $^W;
+   $^W          = 1;
+
+my $mech;
+
+warnings_are
+    {
+        $mech = WWW::Mechanize::Cached->new(
+            cache => $cache,
+        );
+    }
+    [ ],
+    "No warnings for accepted 'cache' parameter";
+
+for my $boolean_attribute (
+    qw(
+        is_cached
+        positive_cache
+        ref_in_cach_key
+        _verbose_dwarn
+        cache_undef_content_length
+        cache_zero_content_length
+        cache_mismatch_content_length
+    )
+) {
+    warnings_are
+        {
+            $mech = WWW::Mechanize::Cached->new(
+                $boolean_attribute => 1,
+            );
+        }
+        [ ],
+        "No warnings for accepted '$boolean_attribute' parameter";
+}
+
+warnings_exist
+    {
+        $mech = WWW::Mechanize::Cached->new(
+            not_my_argument => 1,
+        );
+    }
+    qr/not_my_argument/,
+    'Unrecognized arguments passed through to WWW::Mechanize';
+
+
+# Put this back
+$^W = $was_warning;
+
+done_testing();

